docs/source/manim_directive.py

```diff
@@ -0,0 +1,242 @@
+
+import jinja2
+import os
+from os.path import relpath
+
+import shutil
+
+
+class ManimDirective(Directive):
+    r"""The ``.. manim::`` directive.
+
+    See the module docstring for documentation.
+    """
+    has_content = True
+    required_arguments = 1
+    optional_arguments = 0
+    option_spec = {
+        "display_source": bool,
+        "quality": lambda arg: directives.choice(
+            arg, ("low", "medium", "high", "fourk")
+        ),
+        "save_as_gif": bool,
+        "save_last_frame": bool,
+    }
+    final_argument_whitespace = True
+
+    def run(self):
+        clsname = self.arguments[0]
+
+        display_source = "display_source" in self.options
+        save_as_gif = "save_as_gif" in self.options
+        save_last_frame = "save_last_frame" in self.options
+        assert not (save_as_gif and save_last_frame)
+
+        frame_rate = 60
+        pixel_height = 1080
+        pixel_width = 1920
+
+        if "quality" in self.options:
+            quality = self.options["quality"]
+            if quality == "low":
+                pixel_height = 480
+                pixel_width = 854
+                frame_rate = 15
+            elif quality == "medium":
+                pixel_height = 720
+                pixel_width = 1280
+                frame_rate = 30
+            elif quality == "high":
+                pixel_height = 1440
+                pixel_width = 2560
+                frame_rate = 60
+            elif quality == "fourk":
+                pixel_height = 2160
+                pixel_width = 3840
+                frame_rate = 60
+
+        qualitydir = f"{pixel_height}p{frame_rate}"
+
+        state_machine = self.state_machine
+        document = state_machine.document
+
+        source_file_name = document.attributes["source"]
+        source_rel_name = relpath(source_file_name, setup.confdir)
+        source_rel_dir = os.path.dirname(source_rel_name)
+        while source_rel_dir.startswith(os.path.sep):
+            source_rel_dir = source_rel_dir[1:]
+
+        source_dir = os.path.abspath(
+            os.path.join(setup.app.builder.srcdir, source_rel_dir)
+        )
+
+        dest_dir = os.path.abspath(
+            os.path.join(setup.app.builder.outdir, source_rel_dir)
+        )
+        if not os.path.exists(dest_dir):
+            os.makedirs(dest_dir)
+
+        source_block = [
+            ".. code-block:: python",
+            "",
+            *["    " + line for line in self.content],
+        ]
+        source_block = "\n".join(source_block)
+
+        file_writer_config_code = [
+            f'config["frame_rate"] = {frame_rate}',
+            f'config["pixel_height"] = {pixel_height}',
+            f'config["pixel_width"] = {pixel_width}',
+            'file_writer_config["media_dir"] = "./source/media"',
+            'file_writer_config["images_dir"] = "./source/media/images"',
+            'file_writer_config["video_dir"] = "./source/media/videos"',
```

Comment:
Why not use `os` module path function?
like `os.path.join('source','media')`?

Reply:
Good catch on all of these! I'll take care of it.

Reply:
Taken care of via 2dc15e84.